programs/drift/src/state/lp_pool.rs

```diff
@@ -0,0 +1,1462 @@
+                None,
+                None,
+                0,
+            )?
+        };
+        let in_fee_pct = in_fee_pct.safe_add(out_fee_pct)?;
+        let in_fee_amount = in_amount
+            .cast::<i128>()?
+            .safe_mul(in_fee_pct)?
+            .safe_div(PERCENTAGE_PRECISION_I128)?;
+
+        let in_amount_less_fees = in_amount
+            .cast::<i128>()?
+            .safe_sub(in_fee_amount)?
+            .max(0)
+            .cast::<u128>()?;
+
+        let token_precision_denominator = 10_u128.pow(in_spot_market.decimals);
+        let token_amount_usd = in_oracle
+            .price
+            .cast::<u128>()?
+            .safe_mul(in_amount_less_fees)?;
+        let lp_amount = if self.last_aum == 0 {
+            token_amount_usd.safe_div(token_precision_denominator)?
+        } else {
+            msg!("token_amount_usd: {}", token_amount_usd);
+            msg!("dlp_total_supply: {}", dlp_total_supply);
+            msg!("last_aum: {}", self.last_aum);
+            msg!(
+                "token_precision_denominator: {}",
+                token_precision_denominator
+            );
+            token_amount_usd
+                .safe_mul(dlp_total_supply as u128)?
+                .safe_div(self.last_aum.safe_mul(token_precision_denominator)?)?
+        };
+
+        let lp_fee_to_charge_pct = self.get_mint_redeem_fee(now, true)?;
+        let lp_fee_to_charge = lp_amount
+            .safe_mul(lp_fee_to_charge_pct as u128)?
+            .safe_div(PERCENTAGE_PRECISION)?
+            .cast::<i64>()?;
+
+        Ok((
+            lp_amount.cast::<u64>()?,
+            in_amount,
+            lp_fee_to_charge,
+            in_fee_amount,
+        ))
+    }
+
+    /// Calculates the amount of constituent tokens to receive for a given amount of LP tokens to burn
+    /// Returns the mint_amount in lp token precision and fee to charge in constituent mint precision
+    pub fn get_remove_liquidity_amount(
+        &self,
+        now: i64,
+        out_spot_market: &SpotMarket,
+        out_constituent: &Constituent,
+        lp_burn_amount: u64,
+        out_oracle: &OraclePriceData,
+        out_target_weight: i64,
+        dlp_total_supply: u64,
+    ) -> DriftResult<(u64, u128, i64, i128)> {
+        let lp_fee_to_charge_pct = self.get_mint_redeem_fee(now, false)?;
+        let lp_fee_to_charge = lp_burn_amount
+            .cast::<i128>()?
+            .safe_mul(lp_fee_to_charge_pct.cast::<i128>()?)?
+            .safe_div(PERCENTAGE_PRECISION_I128)?
+            .cast::<i64>()?;
+
+        let lp_amount_less_fees = (lp_burn_amount as i128).safe_sub(lp_fee_to_charge as i128)?;
+        msg!("lp_amount_less_fees: {}", lp_amount_less_fees);
+
+        let token_precision_denominator = 10_u128.pow(out_spot_market.decimals);
+
+        // Calculate proportion of LP tokens being burned
+        let proportion = lp_amount_less_fees
+            .cast::<u128>()?
+            .safe_mul(PERCENTAGE_PRECISION)?
```

Comment:
1e12 might be overkill

Reply:
would good to know the cu's here, 1e12 is pretty giant and might increases CUs quite a bit